engine/crates/composition/src/compose/entity_interface.rs

```diff
@@ -0,0 +1,181 @@
+use super::*;
+use crate::composition_ir as ir;
+use std::collections::{BTreeMap, BTreeSet};
+
+pub(crate) fn merge_entity_interface_definitions(
+    ctx: &mut Context<'_>,
+    first: DefinitionWalker<'_>,
+    definitions: &[DefinitionWalker<'_>],
+) {
+    let interface_name = first.name();
+
+    let iface_defs = || definitions.iter().filter(|def| def.kind() == DefinitionKind::Interface);
+    let mut ifaces = iface_defs();
+
+    let Some(iface_def) = ifaces.next() else {
+        ctx.diagnostics.push_fatal(format!(
+            "The entity interface `{}` is not defined as an interface in any subgraph.",
+            interface_name.as_str()
+        ));
+        return;
+    };
+
+    // More than one interface in subgraphs.
+    if ifaces.next().is_some() {
+        let all_implementers: BTreeSet<_> = iface_defs()
+            .flat_map(|iface| {
+                iface
+                    .subgraph()
+                    .interface_implementers(interface_name.id)
+                    .map(|def| def.name().id)
+            })
+            .collect();
+
+        // All subsequent interfaces must have the same implementers.
+        for iface in iface_defs() {
+            let implementers: BTreeSet<_> = iface
+                .subgraph()
+                .interface_implementers(interface_name.id)
+                .map(|def| def.name().id)
+                .collect();
+
+            if implementers != all_implementers {
+                let subgraph_name = iface.subgraph().name().as_str();
+                let iface_name = interface_name.as_str();
+                let implementer_names = all_implementers
+                    .difference(&implementers)
+                    .map(|id| ctx.subgraphs.walk(*id).as_str())
+                    .join(", ");
+                ctx.diagnostics.push_fatal(format!(
+                r#"[{subgraph_name}] Interface type "{iface_name}" has a resolvable key in subgraph "{subgraph_name}" but that subgraph is missing some of the supergraph implementation types of "{iface_name}". Subgraph "{subgraph_name}" should define types {implementer_names}."#
+                ));
+            }
+
+            if iface.is_interface_object() {
+                ctx.diagnostics.push_fatal(format!(
+                    "[{}] The @interfaceObject directive is not valid on interfaces (on `{}`).",
+                    iface.subgraph().name().as_str(),
+                    interface_name.as_str(),
+                ));
+            }
+        }
+    }
+
+    let interface_id = ctx.insert_interface(interface_name);
+
+    let mut fields = BTreeMap::new();
+
+    for field in iface_def.fields() {
+        fields.entry(field.name().id).or_insert_with(|| ir::FieldIr {
```

Comment:
Could probably implement default to shorten this but not critical

Reply:
It's a bit hard because there are required fields where default could be implemented (StringId::MIN for example) but that would always or nearly always be incorrect not to set them.

Reply:
I see, ok